trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,295 @@
+            rolled = tuple(np.roll(cycle, -cycle.index(node)),)
+            break
+
+    return rolled
+
+
+def insert_cycles(pathway, rolled_cycles):
+    '''Inserts already-rolled cycles as a tuple into a path
+    '''
+    path=list(pathway)
+    for rc in rolled_cycles:
+        path.insert(path.index(rc[0]), rc)
+    path_with_cycles = tuple(path)
+    return tuple(path)
+
+
+def get_pathways_with_cycles(pathways, sc):
+    '''For the given list of pathways and simple cycles, returns a set of
+    pathways that could contain 1 or more cycles
+    '''
+    if len(pathways) is 0:
+        return set()
+
+    pathways_with_cycles = set()
+    for path in pathways:
+        rolled_cycles = set()
+        for cycle in sc:
+            rolled_cycle = roll_cycle(path, cycle)
+            if rolled_cycle:
+                rolled_cycles.add(rolled_cycle)
+        
+        # record all the pathways that have cycles, insert single cycle
+        if rolled_cycles:
+            pathways_with_cycles.add(insert_cycles(path, rolled_cycles))
+
+    return pathways_with_cycles
+
+
+def find_paths_with_source(pathways, source):
+    '''returns a subset of pathways that contain a given facility as the source
+    '''
+    if len(pathways) is 0:
+        return set()
+
+    subset_pathways = set([ path for path in pathways if path[0] == source])
+    return subset_pathways
+
+
+def find_paths_with_sink(pathways, sink):
+    '''returns a subset of pathways that contain a given facility as the sink
+    '''
+    if len(pathways) is 0:
+        return set()
+    
+    subset_pathways = set([ path for path in pathways if path[-1] == sink])
+    return subset_pathways
+
+
+def find_paths_containing_all(pathways, facilities):
+    '''returns a subset of pathways that contain all facilities in input list
+    '''
+    if len(pathways) is 0:
+        return set()
+
+    # convert to list if user passed a string or int
+    if type(facilities) == int or type(facilities) == str:
+        facilities = [facilities]
+    
+    # if user passed an empty list, return no pathways
+    if not facilities:
+        return set()
+
+    p = set([path for path in pathways if set(facilities).issubset(path)])
+
+    return p
+
+
+def find_paths_containing_one_of(pathways, facilities):
+    '''Returns a subset of pathways that contain one or more facilities in
+    input list
+    '''
+    # convert to list if user passed a string or int
+    if type(facilities) == int or type(facilities) == str:
+        facilities = [facilities]
+
+    # if user passed an empty list, return no pathways
+    if len(pathways) is 0 or len(facilities) is 0:
+        return set()
+
+    facilities = set(facilities)
+    p = set([path for path in pathways if set(path).intersection(facilities)])
+
+    return p
+
+
+def get_shortest_path(pathways):
+    '''Finds the set of pathways with the shortest number of steps from source to
+    target. Returns a tuple with path and length.
```

Comment:
I can't recall if we discussed this already, and whether I am contradicting earlier advice, but I'm not convinced that these should return both a length & the set of shortest paths.  The length can be easily computed from the paths if there are any, and it simplifies the interface.

Reply:
Took out the return of lengths